uniffi_bindgen/src/bindings/ruby/templates/ErrorTemplate.rb

```diff
@@ -1,50 +1,89 @@
-class RustError < FFI::Struct
-  layout :code,    :int32,
-         :message, :string
+class RustCallStatus < FFI::Struct
+  layout :code,    :int8,
+         :error_buf, RustBuffer
 
   def code
     self[:code]
   end
 
+  def error_buf
+    self[:error_buf]
+  end
+
   def to_s
-    "RustError(code=#{self[:code]}, message=#{self[:message]})"
+    "RustCallStatus(code=#{self[:code]})"
   end
 end
 
-{% for e in ci.iter_error_definitions() %}
-class {{ e.name()|class_name_rb }}
-  {%- for value in e.values() %}
-  {{ value|class_name_rb }} = Class.new StandardError
+# These match the values from the uniffi::rustcalls module
+CALL_SUCCESS = 1
+CALL_ERROR = 2
+CALL_PANIC = 3
+{%- for e in ci.iter_error_definitions() %}
+module {{ e.name()|class_name_rb }}
+  {%- for variant in e.variants() %}
+  class {{ variant.name()|class_name_rb }} < StandardError
+    def initialize({% for field in variant.fields() %}{{ field.name()|var_name_rb }}{% if !loop.last %}, {% endif %}{% endfor %})
+        {%- for field in variant.fields() %}
+        @{{ field.name()|var_name_rb }} = {{ field.name()|var_name_rb }}
+        {%- endfor %}
+        super()
+      end
+    {%- if variant.has_fields() %}
+
+    attr_reader {% for field in variant.fields() %}:{{ field.name() }}{% if !loop.last %}, {% endif %}{% endfor %}
+    {% endif %}
+  end
   {%- endfor %}
+end
+{%- endfor %}
 
-  def self.raise_err(code, message)
-    {%- for value in e.values() %}
-    if code == {{ loop.index }}
-      raise {{ value|class_name_rb }}, message
-    end
-    {% endfor %}
-    raise 'Unknown error code'
+# Map error modules to the RustBuffer method name that reads them
+@error_module_to_reader_method = {
+{%- for e in ci.iter_error_definitions() %}
+{%- let typ=ci.get_type(e.name()).unwrap() %}
+{%- let canonical_type_name = typ.canonical_name()|class_name_rb %}
+  {{ e.name()|class_name_rb }} => :read{{ canonical_type_name }},
+{%- endfor %}
+}
```

Comment:
Would it be a good idea to make `error_module_to_reader_method` private somehow or just flagging it internal like in Python?  I don't know nearly enough about Ruby here.

Reply:
My second attempt is to make the variable name upper case (`ERROR_MODULE_TO_READER_METHOD`).  That makes the code work, although I'm still not sure if it's idomatic Ruby.

Reply:
I had a quick look. Having this mapping defined as a constant `ERROR_MODULE_TO_READER_METHOD` is definitely a good idea and quite idiomatic for ruby as for me. There are ways to mark constants/classes/modules as private but language still allows these to be accessed with simple meta-programming techniques.

I'll spend more time on this later today.
Thanks!

Reply:
I marked few constants as private @bendk, please have a look https://github.com/bendk/uniffi-rs/pull/1

Reply:
@saks that's wonderful, thanks so much!  I just merged them into the branch I'm working on, they'll be in the final product.